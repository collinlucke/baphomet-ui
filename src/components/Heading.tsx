import { useState } from 'react';
import {
  Button,
  ButtonGroup,
  Header,
  mediaQueries
} from '@collinlucke/phantomartist';
import { Link } from 'react-router-dom';
import { useNavigate, useLocation } from 'react-router-dom';
import { PlusSignIcon, Menu01Icon, Cancel01Icon } from 'hugeicons-react';
import { isAuthenticatedVar, isMobileAndLandscapeVar } from '../reactiveVars';
import { useReactiveVar } from '@apollo/client';
import { CSSObject } from '@emotion/react';

type HeadingProps = {
  setShowLoginModal: (show: boolean) => void;
  setShowSignupModal: (show: boolean) => void;
};

export const Heading: React.FC<HeadingProps> = ({
  setShowLoginModal,
  setShowSignupModal
}) => {
  const navigate = useNavigate();
  const location = useLocation();
  console.log(location);
  const isAuthenticated = useReactiveVar(isAuthenticatedVar);
  const isMobileAndLandscape = useReactiveVar(isMobileAndLandscapeVar);
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
  const [isFloatingMenuOpen, setIsFloatingMenuOpen] = useState(false);
  const user = localStorage.getItem('baphomet-user')
    ? JSON.parse(localStorage.getItem('baphomet-user') || '{}')
    : null;

  const handleAddMovies = () => {
    setIsMobileMenuOpen(false); // Close menu after navigation
  };

  const openSignupModal = () => {
    setShowSignupModal(true);
    setIsMobileMenuOpen(false); // Close menu after action
  };

  const openLoginModal = () => {
    setShowLoginModal(true);
    setIsMobileMenuOpen(false); // Close menu after action
  };

  const logOut = () => {
    localStorage.removeItem('baphomet-token');
    localStorage.removeItem('baphomet-user');
    isAuthenticatedVar(false);
    setIsMobileMenuOpen(false); // Close menu after action
  };

  const handleNavigation = (path: string) => {
    navigate(path);
    setIsMobileMenuOpen(false); // Close menu after navigation
  };

  const toggleMobileMenu = () => {
    setIsMobileMenuOpen(!isMobileMenuOpen);
  };

  const toggleFloatingMenu = () => {
    setIsFloatingMenuOpen(!isFloatingMenuOpen);
  };
  // If mobile landscape, render floating hamburger menu instead
  if (isMobileAndLandscape && location.pathname === '/arena') {
    return (
      <>
        {/* Floating Hamburger Button */}
        <div css={baphStyles.floatingMenuButton}>
          <Button
            kind="primary"
            size="small"
            iconOnly
            icon={
              isFloatingMenuOpen ? (
                <Cancel01Icon size={20} />
              ) : (
                <Menu01Icon size={20} />
              )
            }
            onClick={toggleFloatingMenu}
            ariaLabel={isFloatingMenuOpen ? 'Close menu' : 'Open menu'}
          />
        </div>

        {/* Floating Menu Overlay */}
        {isFloatingMenuOpen && (
          <div
            css={baphStyles.floatingMenuOverlay}
            className="floating-menu-overlay"
            onClick={() => setIsFloatingMenuOpen(false)}
          >
            <div
              css={baphStyles.floatingMenu}
              onClick={e => e.stopPropagation()}
            >
              {/* Logo/Title */}
              <div css={baphStyles.floatingMenuHeader}>
                <Link
                  to={'/'}
                  css={baphStyles.floatingLogoLink}
                  onClick={() => setIsFloatingMenuOpen(false)}
                >
                  <h2 css={baphStyles.floatingTitle}>Baphomet</h2>
                  <img
                    src="/baphy-favicon.png"
                    alt="Baphomet logo"
                    css={baphStyles.floatingFavicon}
                  />
                </Link>
              </div>

              {/* Navigation */}
              <nav css={baphStyles.floatingNav}>
                <div css={baphStyles.floatingNavSection}>
                  <Button
                    size="medium"
                    kind="ghostOnDark"
                    onClick={() => handleNavigation('/arena')}
                    className={{ button: baphStyles.floatingNavButton }}
                  >
                    Arena
                  </Button>
                  <Button
                    size="medium"
                    kind="ghostOnDark"
                    onClick={() => handleNavigation('/leaderboards')}
                    disabled
                    className={{ button: baphStyles.floatingNavButton }}
                  >
                    Leader Boards
                  </Button>
                  <Button
                    size="medium"
                    kind="ghostOnDark"
                    onClick={() => handleNavigation('/all-movies')}
                    className={{ button: baphStyles.floatingNavButton }}
                  >
                    All Movies
                  </Button>
                </div>

                {isAuthenticated && user.role === 'admin' && (
                  <div css={baphStyles.floatingNavSection}>
                    <Button
                      size="medium"
                      kind="secondary"
                      onClick={() => {
                        handleNavigation('/add-movies');
                        handleAddMovies();
                      }}
                      icon={<PlusSignIcon size={20} strokeWidth={'3px'} />}
                      className={{ button: baphStyles.floatingNavButton }}
                    >
                      Add new movie
                    </Button>
                    <Button
                      kind="primary"
                      onClick={logOut}
                      size="medium"
                      className={{ button: baphStyles.floatingNavButton }}
                    >
                      Log out
                    </Button>
                  </div>
                )}

                {!isAuthenticated && (
                  <div css={baphStyles.floatingNavSection}>
                    <Button
                      onClick={openSignupModal}
                      kind="secondary"
                      size="medium"
                      className={{ button: baphStyles.floatingNavButton }}
                    >
                      Sign Up
                    </Button>
                    <Button
                      onClick={openLoginModal}
                      kind="primary"
                      size="medium"
                      className={{ button: baphStyles.floatingNavButton }}
                    >
                      Log in
                    </Button>
                  </div>
                )}
              </nav>
            </div>
          </div>
        )}
      </>
    );
  }

  // Regular header for non-mobile-landscape devices

  return (
    <Header>
      {[
        <div key="header-content" css={baphStyles.headerContent}>
          <Link
            to={'/'}
            data-testid="home-link"
            css={baphStyles.logoLink}
            aria-label="Baphomet - Go to homepage"
          >
            <h1 css={baphStyles.title}>Baphomet</h1>
            <img
              src="/baphy-favicon.png"
              alt="Baphomet logo"
              css={baphStyles.favicon}
              role="presentation"
              aria-hidden="true"
            />
          </Link>

          {/* Desktop Navigation - hidden on mobile via CSS */}
          <div css={baphStyles.rightSideContent}>
            <nav
              css={baphStyles.nav}
              role="navigation"
              aria-label="Main navigation"
            >
              <ButtonGroup ariaLabel="Navigation buttons">
                <Link to="/arena">
                  <Button
                    size="small"
                    kind="ghostOnDark"
                    onClick={() => handleNavigation('/arena')}
                    ariaLabel="Go to Arena page"
                  >
                    Arena
                  </Button>
                </Link>
                {/* <Link to="/leaderboards" > */}
                <Button
                  size="small"
                  kind="ghostOnDark"
                  ariaLabel="Go to Leader Boards page"
                  disabled
                >
                  Leader Boards
                </Button>
                {/* </Link> */}
                <Link to="/all-movies">
                  <Button
                    size="small"
                    kind="ghostOnDark"
                    ariaLabel="Go to All Movies page"
                  >
                    All Movies
                  </Button>
                </Link>
              </ButtonGroup>
            </nav>

            {isAuthenticated && user.role === 'admin' && (
              <ButtonGroup ariaLabel="User actions">
                <Link to="/add-movies">
                  <Button
                    dataTestId="add-new-movie-button"
                    size="small"
                    icon={<PlusSignIcon size={17} strokeWidth={'3px'} />}
                    kind="secondary"
                    ariaLabel="Add new movie"
                    onClick={handleAddMovies}
                  >
                    Add new movie
                  </Button>
                </Link>
                <Button
                  kind="primary"
                  onClick={logOut}
                  size="small"
                  ariaLabel="Sign out of your account"
                  dataTestId="logout-button"
                >
                  Log out
                </Button>
              </ButtonGroup>
            )}

            {!isAuthenticated && (
              <>
                <ButtonGroup ariaLabel="Authentication options">
                  <Button
                    onClick={openSignupModal}
                    kind="secondary"
                    size="small"
                    ariaLabel="Open sign up form"
                    ariaDescribedBy="signup-help"
                    dataTestId="signup-button"
                  >
                    Sign Up
                  </Button>
                  <Button
                    onClick={openLoginModal}
                    kind="primary"
                    size="small"
                    ariaLabel="Open log in form"
                    ariaDescribedBy="login-help"
                    dataTestId="login-button"
                  >
                    Log in
                  </Button>
                </ButtonGroup>
                <span id="signup-help" css={baphStyles.srOnly}>
                  Create a new account to add and rate movies
                </span>
                <span id="login-help" css={baphStyles.srOnly}>
                  Sign in to your existing account
                </span>
              </>
            )}
          </div>

          {/* Mobile Controls - shown only on mobile via CSS */}
          <div css={baphStyles.mobileControls}>
            <Button
              kind="ghostOnDark"
              size="small"
              iconOnly
              icon={
                isMobileMenuOpen ? (
                  <Cancel01Icon size={24} />
                ) : (
                  <Menu01Icon size={24} />
                )
              }
              onClick={toggleMobileMenu}
              ariaLabel={isMobileMenuOpen ? 'Close menu' : 'Open menu'}
            />
          </div>

          {/* Mobile Menu Overlay - shown only when menu is open */}
          {isMobileMenuOpen && (
            <div
              css={baphStyles.mobileMenuOverlay}
              onClick={() => setIsMobileMenuOpen(false)}
            >
              <div
                css={baphStyles.mobileMenu}
                onClick={e => e.stopPropagation()}
              >
                <nav
                  css={baphStyles.mobileNav}
                  role="navigation"
                  aria-label="Mobile navigation"
                >
                  <div css={baphStyles.mobileNavSection}>
                    <Button
                      size="medium"
                      kind="ghostOnDark"
                      onClick={() => handleNavigation('/arena')}
                      className={{ button: baphStyles.mobileNavButton }}
                    >
                      Arena
                    </Button>
                    <Button
                      size="medium"
                      kind="ghostOnDark"
                      onClick={() => handleNavigation('/leaderboards')}
                      disabled
                      className={{ button: baphStyles.mobileNavButton }}
                    >
                      Leader Boards
                    </Button>
                    <Button
                      size="medium"
                      kind="ghostOnDark"
                      onClick={() => handleNavigation('/all-movies')}
                      className={{ button: baphStyles.mobileNavButton }}
                    >
                      All Movies
                    </Button>
                  </div>

                  {isAuthenticated && user.role === 'admin' && (
                    <div css={baphStyles.mobileNavSection}>
                      <Link to="/add-movies">
                        <Button
                          size="medium"
                          kind="secondary"
                          onClick={handleAddMovies}
                          icon={<PlusSignIcon size={20} strokeWidth={'3px'} />}
                          className={{ button: baphStyles.mobileNavButton }}
                        >
                          Add new movie
                        </Button>
                      </Link>
                      <Button
                        kind="primary"
                        onClick={logOut}
                        size="medium"
                        className={{ button: baphStyles.mobileNavButton }}
                      >
                        Log out
                      </Button>
                    </div>
                  )}

                  {!isAuthenticated && (
                    <div css={baphStyles.mobileNavSection}>
                      <Button
                        onClick={openSignupModal}
                        kind="secondary"
                        size="medium"
                        className={{ button: baphStyles.mobileNavButton }}
                      >
                        Sign Up
                      </Button>
                      <Button
                        onClick={openLoginModal}
                        kind="primary"
                        size="medium"
                        className={{ button: baphStyles.mobileNavButton }}
                      >
                        Log in
                      </Button>
                    </div>
                  )}
                </nav>
              </div>
            </div>
          )}
        </div>
      ]}
    </Header>
  );
};

// Define base styles with mobile responsiveness
const baphStyles: { [key: string]: CSSObject } = {
  headerContent: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
    padding: '0 15px',
    [mediaQueries.minWidth.sm]: {
      padding: '0 20px'
    }
  },
  logoLink: {
    display: 'flex',
    alignItems: 'center',
    gap: '6px',
    textDecoration: 'none',
    [mediaQueries.minWidth.sm]: {
      gap: '8px'
    },
    [mediaQueries.minWidth.md]: {
      gap: '10px'
    },
    [mediaQueries.minWidth.lg]: {
      gap: '12px'
    }
  },
  title: {
    margin: 0,
    fontSize: '1.2rem',
    [mediaQueries.minWidth.sm]: {
      fontSize: '1.4rem'
    },
    [mediaQueries.minWidth.md]: {
      fontSize: '1.6rem'
    },
    [mediaQueries.minWidth.lg]: {
      fontSize: '2rem'
    }
  },
  favicon: {
    width: '32px',
    height: '32px',
    objectFit: 'contain' as const,
    [mediaQueries.minWidth.sm]: {
      width: '40px',
      height: '40px'
    },
    [mediaQueries.minWidth.md]: {
      width: '48px',
      height: '48px'
    },
    [mediaQueries.minWidth.lg]: {
      width: '64px',
      height: '64px'
    }
  },
  rightSideContent: {
    display: 'none', // Hidden on mobile by default
    alignItems: 'center',
    gap: '15px',
    flex: 1,
    justifyContent: 'flex-end',
    [mediaQueries.minWidth.md]: {
      display: 'flex', // Show on desktop (640px+)
      gap: '20px'
    }
  },
  nav: {
    display: 'flex',
    alignItems: 'center',
    gap: '15px',
    [mediaQueries.minWidth.md]: {
      gap: '20px'
    }
  },
  mobileControls: {
    display: 'flex', // Show on mobile by default
    alignItems: 'center',
    [mediaQueries.minWidth.md]: {
      display: 'none' // Hide on desktop (640px+)
    }
  },
  mobileMenuOverlay: {
    position: 'fixed' as const,
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.8)',
    zIndex: 1000,
    display: 'flex',
    justifyContent: 'flex-end'
  },
  mobileMenu: {
    backgroundColor: '#0B1828', // Using primary color
    width: '280px',
    height: '100%',
    padding: '20px',
    boxShadow: '-2px 0 8px rgba(0, 0, 0, 0.3)',
    overflowY: 'auto' as const
  },
  mobileNav: {
    display: 'flex',
    flexDirection: 'column' as const,
    gap: '20px',
    paddingTop: '60px' // Account for header height
  },
  mobileNavSection: {
    display: 'flex',
    flexDirection: 'column' as const,
    gap: '12px',
    borderBottom: '1px solid rgba(255, 255, 255, 0.1)',
    paddingBottom: '20px',
    '&:last-child': {
      borderBottom: 'none',
      paddingBottom: 0
    }
  },
  mobileNavButton: {
    width: '100%',
    justifyContent: 'flex-start'
  },
  srOnly: {
    position: 'absolute' as const,
    visibility: 'hidden' as const
  },

  // Floating menu styles for mobile landscape
  floatingMenuButton: {
    position: 'fixed' as const,
    top: '10px',
    right: '10px',
    zIndex: 1001,
    backgroundColor: 'rgba(11, 24, 40, 0.9)', // Semi-transparent primary color
    borderRadius: '50%',
    padding: '8px',
    boxShadow: '0 4px 12px rgba(0, 0, 0, 0.3)'
  },
  floatingMenuOverlay: {
    position: 'fixed' as const,
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.8)',
    zIndex: 1000,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center'
  },
  floatingMenu: {
    backgroundColor: '#0B1828',
    borderRadius: '12px',
    padding: '20px',
    maxWidth: '320px',
    width: '90%',
    maxHeight: '80vh',
    overflowY: 'auto' as const,
    boxShadow: '0 8px 32px rgba(0, 0, 0, 0.5)',
    border: '1px solid rgba(255, 255, 255, 0.1)'
  },
  floatingMenuHeader: {
    display: 'flex',
    justifyContent: 'center',
    marginBottom: '20px',
    paddingBottom: '15px',
    borderBottom: '1px solid rgba(255, 255, 255, 0.1)'
  },
  floatingLogoLink: {
    display: 'flex',
    alignItems: 'center',
    gap: '8px',
    textDecoration: 'none',
    color: 'inherit'
  },
  floatingTitle: {
    margin: 0,
    fontSize: '1.4rem',
    color: '#FFFFFF'
  },
  floatingFavicon: {
    width: '32px',
    height: '32px',
    objectFit: 'contain' as const
  },
  floatingNav: {
    display: 'flex',
    flexDirection: 'column' as const,
    gap: '15px'
  },
  floatingNavSection: {
    display: 'flex',
    flexDirection: 'column' as const,
    gap: '10px',
    borderBottom: '1px solid rgba(255, 255, 255, 0.1)',
    paddingBottom: '15px',
    '&:last-child': {
      borderBottom: 'none',
      paddingBottom: 0
    }
  },
  floatingNavButton: {
    width: '100%',
    justifyContent: 'flex-start'
  }
};
