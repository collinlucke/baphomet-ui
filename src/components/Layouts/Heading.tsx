import { Header, mediaQueries, Button } from 'phantomartist';
import { Menu01Icon } from 'hugeicons-react';
import { LogoLink } from './LogoLink';
import {
  showSlideOutMenuVar,
  isLargeScreenVar,
  isMobileVar,
  isSmallOrMobileVar,
  isAuthenticatedVar
} from '../../reactiveVars';
import { useReactiveVar } from '@apollo/client/react';
import { CSSObject } from '@emotion/react';
import { MainNavLinks } from '../Navs/MainNavLinks';
import { SlideOutNav } from '../Navs/SlideOutNav';
import { AuthUserAvatarButton } from '../Navs/AuthUserAvatarButton';
import { UnauthorizedButtons } from '../Navs/UnauthorizedButtons';

export const Heading: React.FC = () => {
  const isLargeScreen = useReactiveVar(isLargeScreenVar);
  const isMobile = useReactiveVar(isMobileVar);
  const isSmallOrMobile = useReactiveVar(isSmallOrMobileVar);
  const isAuthenticated = useReactiveVar(isAuthenticatedVar);
  const user = localStorage.getItem('baphomet-user')
    ? JSON.parse(localStorage.getItem('baphomet-user') || '{}')
    : null;

  const openSlideOutMenuHandler = () => {
    if (isSmallOrMobile) {
      showSlideOutMenuVar(true);
    }
  };

  const MenuIconButton = () => (
    <Button
      variant="ghostOnDark"
      size="small"
      iconOnly
      icon={<Menu01Icon size={24} />}
      onClick={openSlideOutMenuHandler}
      ariaLabel={'Open menu'}
      className={{ button: baphStyles.openButton }}
    />
  );

  return (
    <Header className={{ header: getHeaderStyes(isSmallOrMobile) }}>
      <div key="header-content" css={baphStyles.headerContent}>
        <nav css={baphStyles.nav}>
          <LogoLink />

          <div css={baphStyles.rightContent}>
            {isMobile || !isLargeScreen ? (
              <div css={baphStyles.rightContent}>
                <MenuIconButton />
                <SlideOutNav displayName={user?.displayName} />
              </div>
            ) : (
              <>
                <MainNavLinks />
                {isAuthenticated ? (
                  <AuthUserAvatarButton displayName={user?.displayName} />
                ) : (
                  <UnauthorizedButtons />
                )}
              </>
            )}
          </div>
        </nav>
      </div>
    </Header>
  );
};

const getHeaderStyes = (isSmallOrMobile: boolean) => ({
  height: isSmallOrMobile ? '65px' : '75px'
});
const baphStyles: { [key: string]: CSSObject } = {
  header: {
    maxHeight: '55px'
  },
  headerContent: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
    padding: '0 15px',
    [mediaQueries.minWidth.sm]: {
      padding: '0 20px'
    }
  },
  nav: {
    display: 'flex',
    justifyContent: 'space-between',
    gap: '20px',
    width: '100%'
  },
  openButton: {
    padding: '16px',
    zIndex: 10,
    '&:hover': { boxShadow: 'none' }
  },
  rightContent: {
    display: 'flex',
    justifyContent: 'space-between',
    gap: '20px',
    alignItems: 'center'
  }
};
