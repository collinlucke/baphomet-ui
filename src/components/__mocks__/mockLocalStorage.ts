export const mockLocalStorage = () => {
  localStorage.clear();
  const mockLocalStorage: Record<string, unknown> = {};
  global.localStorage = {
    length: 0,
    key: (index: number) => {
      const keys = Object.keys(mockLocalStorage);
      return keys[index] || null;
    },
    setItem: (key, value) => {
      mockLocalStorage[key] = value;
    },
    getItem: key => {
      const value = mockLocalStorage[key];
      return typeof value === 'string' ? value : null;
    },
    removeItem: key => {
      delete mockLocalStorage[key];
    },
    clear: () => {
      for (let key in mockLocalStorage) {
        delete mockLocalStorage[key];
      }
    }
  };
  return () => {};
};
