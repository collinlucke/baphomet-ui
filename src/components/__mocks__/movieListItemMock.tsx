export const createMovieListItemMock = () => ({
  MovieListItem: ({
    movie,
    openMovieDetails
  }: {
    movie: {
      id: string;
      title: string;
      winningPercentage: number;
      tmdbId: string;
    };
    openMovieDetails: (tmdbId: string) => void;
  }) => (
    <li data-testid={`movie-item-${movie.id}`}>
      <span>{movie.title}</span>
      <span data-testid={`winning-percentage-${movie.id}`}>
        {movie.winningPercentage}%
      </span>
      <button
        data-testid={`details-button-${movie.id}`}
        onClick={() => openMovieDetails(movie.tmdbId)}
      >
        Details
      </button>
    </li>
  )
});
