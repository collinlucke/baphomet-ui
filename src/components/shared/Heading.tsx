import { Header, Button, InnerWidth, Block } from '@collinlucke/phantomartist';
import { Link } from 'react-router-dom';
import { useNavigate } from 'react-router-dom';
import { useScreenSize } from '../../hooks/useScreenSize';
import { PlusSignIcon } from 'hugeicons-react';

export const Heading: React.FC = () => {
  const navigate = useNavigate();
  const screenSize = useScreenSize();
  const navToCreate = () => {
    navigate('/create');
  };

  return (
    <Header dataTestId="main-page-heading">
      <Block>
        <InnerWidth className={{ innerWidth: baphStyles.innerWidth }}>
          <h1>
            <Link to={'/'} data-testid="home-link">
              Baphomet
            </Link>
          </h1>
          <div>
            <Button
              dataTestId="add-new-movie-button"
              size={screenSize}
              onClick={navToCreate}
              icon={
                <PlusSignIcon
                  size={screenSize === 'small' ? 20 : 17}
                  strokeWidth={'3px'}
                />
              }
              iconOnly={screenSize === 'small'}
              className={{ button: baphStyles.button }}
            >
              {screenSize === 'small' ? '' : <>Add new movie</>}
            </Button>
          </div>
        </InnerWidth>
      </Block>
    </Header>
  );
};

const baphStyles = {
  innerWidth: {
    flexDirection: 'row' as 'row',
    alignItems: 'end'
  },
  button: {
    position: 'relative' as 'relative'
  }
};
