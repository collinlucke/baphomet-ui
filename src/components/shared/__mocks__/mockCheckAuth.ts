import { CHECK_AUTH } from '../../../api/queries';

export default {
  request: {
    query: CHECK_AUTH,
    variables: {
      token: null
    }
  },
  error: {
    name: 'ApolloError',
    graphQLErrors: [],
    protocolErrors: [],
    clientErrors: [],
    networkError: {
      name: 'ServerError',
      response: {},
      statusCode: 401,
      result: 'Unauthorized'
    },
    message: 'Response not successful: Received status code 401',
    cause: {
      name: 'ServerError',
      response: {},
      statusCode: 401,
      result: 'Unauthorized'
    },
    title: 'checkAuth'
  }
};
