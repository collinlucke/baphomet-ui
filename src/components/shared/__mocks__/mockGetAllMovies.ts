import { GET_ALL_MOVIES } from '../../../api/queries';

export default {
  request: {
    query: GET_ALL_MOVIES,
    variables: {
      limit: 100,
      searchTerm: ''
    }
  },
  result: {
    data: {
      allMovies: [
        {
          fullplot: '',
          poster: null,
          title: 'A new title',
          releaseDate: '1111-03-21',
          rated: null,
          id: '672e8b1d6ec1b1001a9e3c11',
          __typename: 'Movie'
        },
        {
          fullplot:
            'Two co-dependent high school seniors are forced to deal with separation anxiety after their plan to stage a booze-soaked party goes awry.',
          poster:
            'https://m.media-amazon.com/images/M/MV5BMTc0NjIyMjA2OF5BMl5BanBnXkFtZTcwMzIxNDE1MQ@@._V1_SY1000_SX677_AL_.jpg',
          title: 'Superbad',
          releaseDate: '2007-08-17',
          rated: 'R',
          id: '66fdd70418e00a520662b994',
          __typename: 'Movie'
        },
        {
          fullplot: 'new in the plot',
          poster:
            'https://m.media-amazon.com/images/M/MV5BNDI2MzBhNzgtOWYyOS00NDM2LWE0OGYtOGQ0M2FjMTI2NTllXkEyXkFqcGc@._V1_.jpg',
          title: 'The Hangover',
          releaseDate: '2009-06-07',
          rated: 'R',
          id: '66fddf0318e00a520662b997',
          __typename: 'Movie'
        },
        {
          fullplot: '',
          poster: null,
          title: 'added even if not ready',
          releaseDate: null,
          rated: null,
          id: '67355695525d47e2604773b2',
          __typename: 'Movie'
        },
        {
          fullplot: '',
          poster: null,
          title: 'dwdwaqdea',
          releaseDate: null,
          rated: null,
          id: '672e8bedb2a88166073b558a',
          __typename: 'Movie'
        },
        {
          fullplot: '',
          poster: null,
          title: 'fou uyiyhuikuf',
          releaseDate: null,
          rated: null,
          id: '672e8bbfb2a88166073b5589',
          __typename: 'Movie'
        },
        {
          fullplot: '',
          poster: null,
          title: 'the newest of titles;i',
          releaseDate: null,
          rated: null,
          id: '672e8c9579f0d637d1a17060',
          __typename: 'Movie'
        }
      ]
    }
  }
};
