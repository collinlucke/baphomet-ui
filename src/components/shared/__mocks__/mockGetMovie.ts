import { GET_MOVIE } from '../../../api/queries';

export default {
  request: {
    query: GET_MOVIE,
    variables: {
      id: 'test-movie-id'
    }
  },
  result: {
    data: {
      movie: {
        fullplot: 'The is the movie of the little unit test that could',
        poster: 'www.there-is-no-poster.baph',
        title: 'Test Movie Title',
        releaseDate: '06/29/1984',
        rated: 'G',
        id: 'test-movie-id',
        __typename: 'Movie'
      }
    }
  }
};
