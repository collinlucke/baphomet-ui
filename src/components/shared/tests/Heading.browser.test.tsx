import '@testing-library/jest-dom';
import { expect, describe, it, beforeEach } from 'vitest';
import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { mockLocalStorage } from '../__mocks__/mockLocalStorage';
import { Heading } from '../Heading';
import {
  BrowserRouter,
  createBrowserRouter,
  RouterProvider,
  createMemoryRouter
} from 'react-router-dom';
import routes from '../../../routes';
import { MockedProvider } from '@apollo/client/testing';
import mockGetAllMovies from '../__mocks__/mockGetAllMovies';
import mockCheckAuth from '../__mocks__/mockCheckAuth';
import mockGetMovie from '../__mocks__/mockGetMovie';

beforeEach(() => mockLocalStorage());

describe('Heading', () => {
  describe(`'Baphomet' title`, () => {
    it(`renders correctly as a link`, async () => {
      render(
        <BrowserRouter>
          <Heading />
        </BrowserRouter>
      );

      expect(await screen.findByTestId('home-link')).toBeVisible();
    });

    it(`when clicked, it automatically redirects to /movielist if 'beenHereBefore' HAS been set`, async () => {
      // Pretend that we've been to the site before
      localStorage.setItem('beenHereBefore', 'yup');

      const router = createMemoryRouter(routes, {
        initialEntries: ['/view/test-movie-id']
      });

      render(
        <MockedProvider
          mocks={[mockGetAllMovies, mockGetMovie, mockCheckAuth]}
          addTypename={false}
        >
          <RouterProvider router={router} />
        </MockedProvider>
      );

      // Start from /view page so we have a unauthenticated and non-movielist starting point
      expect(screen.getByTestId('movie-editor-form')).toBeVisible();

      // Finds and clicks the home link
      userEvent.click(screen.getByTestId('home-link'));

      // Automatically redirects to the <MoviewListPage>
      expect(await screen.findByTestId('movie-list')).toBeVisible();
    });

    it(`when clicked, goes to the <WelcomePage> if 'beenHereBefore' HAS NOT been set`, async () => {
      const router = createMemoryRouter(routes, {
        initialEntries: ['/view/test-movie-id']
      });

      render(
        <MockedProvider
          mocks={[mockCheckAuth, mockGetMovie]}
          addTypename={false}
        >
          <RouterProvider router={router} />
        </MockedProvider>
      );
      // Make sure we start from /view
      expect(screen.getByTestId('movie-editor-form')).toBeVisible();

      // Finds and clicks the home link in the Heading

      await act(async () => {
        userEvent.click(screen.getByTestId('home-link'));
      });

      // Automatically redirects to the <WelcomePage/>
      expect(await screen.findByTestId('welcome-page')).toBeVisible();
    });
  });

  describe(`'Add new' button`, () => {
    it(`renders the button correctly`, async () => {
      render(
        <BrowserRouter>
          <Heading />
        </BrowserRouter>
      );

      const buttonElem = await screen.findByTestId('add-new-movie-button');

      expect(buttonElem.tagName).toBe('BUTTON');
    });

    it(`redirects to 401 error modal when 'Add new movie' button is clicked`, async () => {
      localStorage.setItem('beenHereBefore', 'yup');

      const router = createBrowserRouter(routes);

      await act(async () => {
        render(
          <MockedProvider
            mocks={[mockGetAllMovies, mockCheckAuth, mockCheckAuth]}
          >
            <RouterProvider router={router} />
          </MockedProvider>
        );
      });

      // Make sure we start out at the <MovieList>
      expect(await screen.findByTestId('movie-list')).toBeVisible();

      const buttonElem = screen.getByTestId('add-new-movie-button');
      await userEvent.click(buttonElem);

      // Confirm Heading is gone as error modals don't allow it
      await waitFor(() => {
        expect(screen.queryByTestId('main-page-heading')).toBeNull();
      });

      expect(await screen.findByTestId('error-modal-401')).toBeVisible();
    });
  });

  it('matches the snapshot', async () => {
    const { container } = render(
      <BrowserRouter>
        <Heading />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
