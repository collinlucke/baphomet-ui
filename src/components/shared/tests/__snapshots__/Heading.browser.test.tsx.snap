// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Heading > matches the snapshot 1`] = `
<div>
  <header
    class="pa-header css-o2h1fg"
    data-testid="main-page-heading"
  >
    <div
      class="pa-block css-xfdfty"
    >
      <div
        class="pa-inner-width css-l3e7vp"
      >
        <h1>
          <a
            data-testid="home-link"
            href="/"
          >
            Baphomet
          </a>
        </h1>
        <div>
          <button
            class="pa-button primary large css-z4rkrz"
            data-testid="add-new-movie-button"
          >
            <svg
              class=""
              color="currentColor"
              fill="none"
              height="17"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="3px"
              viewBox="0 0 24 24"
              width="17"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 4V20M20 12H4"
                stroke="currentColor"
              />
            </svg>
            Add new movie
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;
